server/src/main/java/org/opensearch/gateway/AsyncShardBatchFetch.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.gateway;
+
+import org.apache.logging.log4j.Logger;
+import org.opensearch.action.support.nodes.BaseNodeResponse;
+import org.opensearch.action.support.nodes.BaseNodesResponse;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.cluster.node.DiscoveryNodes;
+import org.opensearch.core.index.shard.ShardId;
+import org.opensearch.indices.store.ShardAttributes;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * Implementation of AsyncShardFetchAbstract with batching support.
+ * cache will be created using ShardBatchCache class as that can handle the caching strategy correctly for a
+ * batch of shards. Other necessary functions are also stored so cache can store or get the data for both primary
+ * and replicas.
+ *
+ * @param <T> Response type of the transport action.
+ * @param <V> Data type of shard level response.
+ */
+public abstract class AsyncShardBatchFetch<T extends BaseNodeResponse, V extends BaseShardResponse> extends AsyncShardFetch<T> {
+
+    private final Consumer<ShardId> removeShardFromBatch;
+    private final List<ShardId> failedShards;
+
+    @SuppressWarnings("unchecked")
+    AsyncShardBatchFetch(
+        Logger logger,
+        String type,
+        Map<ShardId, ShardAttributes> shardAttributesMap,
+        AsyncShardFetch.Lister<? extends BaseNodesResponse<T>, T> action,
+        String batchId,
+        Class<V> clazz,
+        BiFunction<DiscoveryNode, Map<ShardId, V>, T> responseConstructor,
+        Function<T, Map<ShardId, V>> shardsBatchDataGetter,
+        Supplier<V> emptyResponseBuilder,
+        Consumer<ShardId> handleFailedShard
+    ) {
+        super(logger, type, shardAttributesMap, action, batchId);
+        this.removeShardFromBatch = handleFailedShard;
+        this.failedShards = new ArrayList<>();
+        this.cache = new ShardBatchCache<>(
```

Comment:
Right now the challenge with passing cache in super is that we need to pass a method as argument in cache initialization. `this::cleanUpFailedShards`

```
this.cache = new ShardBatchCache<>(
            logger,
            type,
            shardAttributesMap,
            "BatchID=[" + batchId + "]",
            clazz,
            responseGetter,
            shardsBatchDataGetter,
            emptyResponseBuilder,
            this::cleanUpFailedShards
        );
```

I am unable to use **this** before calling super. That's why assigning the cache parameter after calling super. 

Another way would be to pass the cache object without this `this::cleanUpFailedShards` handler and later call a method of AsyncShardFetch to update it. Which won't make sense because this cleanUpFailedShards is only needed for batches and not for normal shard level cache.

So, I'm looking for options to modify it if you have anything in mind.  😃

Reply:
Looks like there is a cyclic dependency being introduced here :)
AsyncShardBatchFetch -> ShardCache -> AsyncShardBatchFetch.

Do we really need a callback from ShardBatchCache given that AsyncShardBatchFetch is using it to maintain the state. `cleanUpFailedShards` looks to be more of a method in AsyncShardFetch just like the existing method `clearCacheForNode`

Reply:
cleanUpFailedShards does 3 things 
1. Remove the shard from ShardsBatch of ShardBatchGateAllocator so that shard can be picked up in next reroute.
2. Remove the shard from shardAttributesMap, so any reroute (even before all nodes finish fetching) doesn't pick this shard in fetch.
3. Add failed shard in a list, so a reroute can be triggered for those shards.

There is one way to avoid passing `this::cleanupFailedShards` in ShardBatchCache constructor.  And handle all 3 functionalities.

If we keep `failedShards` in AsyncShardFetch method fetchData as a local variable only, then `removeFromBatch` function interface can be passed as it is to ShardBatchCache constructor. 

We'll need to keep one more array in `NodeEntry` of **ShardBatchCache**. 
shardFailures [] and keep updating it based on the failures. (It'll be used later to trigger reroute, similar to node failures today)

There is only one challenge, we also need to remove shard from the `shardAttributesMap` as soon as the response is received from transport call. That handler also needs to be added. 

I feel it's just too much of overhead, instead just keeping it protected to be used by child class makes sense to me and it's cleaner. I'm planning to create a setter for `cache` parameter of AsyncShardFetch class, which would be helpful.

Reply:
Setter removed now.